contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -266,41 +304,565 @@ const (
+  - matchNodeLabels:
+      failure-domain.beta.kubernetes.io/zone: us-central1-b
+```
+
+#### Multi-Zonal Example
+This example restricts the volume's primary and failover zones
+to us-central1-a and us-central1-b. Topologies that are incompatible with the
+storage provider parameters will be enforced by the provisioner.
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  Name: multi-zonal-class
+provisioner: kubernetes.io/gce-pd
+parameters:
+  type: pd-standard
+  replication-type: regional-pd
+allowedTopologies:
+  requiredToplogies:
+  - matchNodeLabels:
+      failure-domain.beta.kubernetes.io/zone: us-central1-a
+  - matchNodeLabels:
+      failure-domain.beta.kubernetes.io/zone: us-central1-b
+```
+
+#### Multi Label Example
+This example restricts the volume's topology to nodes that
+have the following labels:
+
+* "zone: us-central1-a" and "rack: rack1" or,
+* "zone: us-central1-b" and "rack: rack1" or,
+* "zone: us-central1-b" and "rack: rack2"
+
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  Name: something-fancy
+provisioner: rack-based-provisioner
+parameters:
+allowedTopologies:
+  requiredTopologies:
+  - matchNodeLabels:
+      zone: us-central1-a
+      rack: rack1
+  - matchNodeLabels:
+      zone: us-central1-b
+      rack: rack1
+  - matchNodeLabels:
+      zone: us-central1-b
+      rack: rack2
+```
+
+### Storage Provider Capacity Reporting
+Some volume types have capacity limitations per topology domain. For example,
+local volumes have limited capacity per node. Provisioners for these volume
+types can report these capacity limitations through a new StorageClass status
+field. It should be noted that provisioners only need to report capacity in the
+StorageClass if the single StorageClass encompasses multiple storage pools with
+different capacities across different topology domains.
+
+```
+type StorageClass struct {
+    ...
+
+    Status *StorageClassStatus
+}
+
+type StorageClassStatus struct {
+    Capacity *StorageClassCapacity
+}
+
+type StorageClassCapacity struct {
+    // Minimum size of a provisioned volume
+    MinVolumeSize resource.Quantity
+
+    // The node topology label keys that the provisioner operates on
+    TopologyKeys []string
+
+    // Capacities specifies the total capacity available
+    // (free and used) for provisioning at each topology value.
+    //
+    // The key of the map is the topology value. It is a flattened,
+    // comma-separated string of label values corresponding with
+    // each label key specified in TopologyKeys in the same order.
+    //
+    // The value of the map is the total capacity for that topology
+    // value.
+    Capacities map[string]resource.Quantity
+}
+```
+
+If no StorageClassCapacity is specified, then there are no capacity
+limitations.  A missing map entry for a topology domain will be treated
+as zero capacity.
+
+A scheduler predicate will use the reported capacity to determine if dynamic
```

Comment:
Also it will scale badly for large clusters.

Reply:
Also it is insufficient to express whether a given volume can actually fit (e.g. could be fragmented).

Reply:
This is similar to how a node today reports total CPU/memory.  It doesn't actually report used vs available, instead the scheduler keeps track of used based on all the Pods that have been scheduled to that node.  This way you avoid the latency issue of waiting for reported values to updated after they are allocated.

So we're trying to follow a similar concept here, except the challenge with volumes is that their topology is not always scoped to a single node, so it doesn't make sense to report as part of node capacity.  The reported capacity would only change if you change the capacity of the storage pool, but not when you allocate/free volumes from it.  I expect the former will happen less frequently than the latter, so the scaling limits should be similar to how we process cpu/memory capacity.

But this does have some limitations to be accurate, like you mentioned:
* The storage pool has to be dedicated to this StorageClass and cannot be shared with other StorageClasses
* Provisioning from the storage pool cannot be prone to fragmentation

Reply:
Here's an alternative idea.  Considering that only local volumes plan to use the capacity reporting api at the moment, and there are some restrictions on how you deploy the storage pools to ensure that this capacity handling is actually accurate, we could try going with a CRD + scheduler extender approach.

The downsides are:
- Deployment is more complicated
- We can't include this into Kubernetes core.  The provisioner is already out of tree, maybe this not an issue.
- The predicate will become best-effort, due to the delay of seeing the PVC provisioning API update in the extender.  The fact that the scheduler round robins on equal priority nodes could help mitigate this, so that the same node is not chosen again for the next Pod.

Advantages are:
- Since it's a CRD specific to local volumes, the API could be simpler and more flexible because we don't need to make it generic to all volume topologies
- You potentially could support scenarios like storage pools being shared across multiple StorageClasses, since it doesn't need to be strictly tied to a StorageClass anymore
- You can encode special behavior/requirements of your specific storage system into the extender logic.
- Keeps StorageClass lean and we don't have to regret not putting a Spec/Status in the beginning

@lichuqiang @jsafrane @thockin wdyt?